include/targets/hip/tunable_kernel.h

```diff
@@ -12,26 +12,50 @@ namespace quda
 {
 
   /**
-      @brief Wrapper around cudaLaunchKernel
-      @param[in] func Device function symbol
-      @param[in] tp TuneParam containing the launch parameters
-      @param[in] arg Host address of argument struct
-      @param[in] stream Stream identifier
-   */
-  qudaError_t qudaLaunchKernel(const void *func, const TuneParam &tp, const qudaStream_t &stream, const void *arg);
+     @brief Wrapper around hipLaunchKernel
+     @param[in] func Device function symbol
+     @param[in] tp TuneParam containing the launch parameters
+     @param[in] arg Host address of argument struct
+     @param[in] stream Stream identifier
+  */
+  qudaError_t qudaLaunchKernel(const kernel_t &kernel, const TuneParam &tp, const qudaStream_t &stream, const void *arg);
+
+  /**
+     @brief Wrapper around hipOccupancyMaxActiveBlocks
+     @param[in] func Device function symbol
+     @param[in] tp TuneParam containing the launch parameters
+  */
+  int qudaOccupancyMaxActiveBlocks(const kernel_t &kernel, const TuneParam &tp);
 
   class TunableKernel : public Tunable
   {
 
   protected:
     QudaFieldLocation location;
 
+    /**
+       @brief Set the maximum number of blocks that can reside on an
+       SM.  This is called when we are autotuning to allow us to work
+       out how many different shared memory over allocations we should
+       use to minimally cover all occupancy variations.
+     */
+    void setMaxActiveBlocks(const kernel_t &kernel, const TuneParam &tp) const
+    {
+      if (activeTuningWarmup() && tuneSharedBytes()) {
+        auto tp2 = tp;
+        setSharedBytes(tp2);
+        // only compute max number blocks when we have no shared memory over subscription
```

Comment:
This logic feels a bit fragile... I understand why this only gets called once (warm-up only happens once), but why does this equality guarantee there's no shared memory oversubscription? Is it because `tp` is at its default setting now, which is `shared_bytes == 0`?